File: app/src/main/java/com/example/ugcssample/drone/GenericUpdateCallback.java
package com.example.ugcssample.drone;

import android.content.Intent;

import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.example.ugcssample.utils.AppScheduledExecutorService;
import com.example.ugcssample.utils.ThreadUtils;

public abstract class GenericUpdateCallback {

    protected static final AppScheduledExecutorService WORKER
            = ThreadUtils.newSingleThreadScheduledExecutor(GenericUpdateCallback.class);

    private final LocalBroadcastManager lbm;

    public GenericUpdateCallback(LocalBroadcastManager lbm) {
        this.lbm = lbm;
    }

    public abstract void setUpKeyListeners();

    public abstract void tearDownKeyListeners();

    protected void sendBroadcast(Intent i) {
        lbm.sendBroadcast(i);
    }

    protected void submitBroadcast(final Intent i) {
        WORKER.submit(new Runnable() {
            @Override
            public void run() {
                sendBroadcast(i);
            }
        });
    }

}

File: app/src/main/java/com/example/ugcssample/drone/GenericUpdateCallback.kt
package com.example.ugcssample.drone

import android.content.Intent
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.example.ugcssample.utils.ThreadUtils

abstract class GenericUpdateCallback(private val lbm: LocalBroadcastManager?) {
    abstract fun setUpKeyListeners()
    abstract fun tearDownKeyListeners()
    protected fun sendBroadcast(i: Intent?) {
        if (i != null) {
            lbm?.sendBroadcast(i)
        }
    }

    protected fun submitBroadcast(i: Intent?) {
        WORKER.submit { sendBroadcast(i) }
    }

    companion object {
        protected val WORKER = ThreadUtils.newSingleThreadScheduledExecutor(
            GenericUpdateCallback::class.java
                                                                           )
    }
}